tools/c7n_azure/c7n_azure/resources/cosmos_db.py

```diff
@@ -52,6 +74,193 @@ class resource_type(ArmResourceManager.resource_type):
+                          " or firewall is blocking access.")
+            else:
+                print(e)
+            return None
+
+        for d in databases:
+            d.update({'c7n:document-endpoint':
+                      parent_resource.get('properties').get('documentEndpoint')})
+
+        return databases
+
+
+@resources.register('cosmosdb-collection')
+class CosmosDBCollection(ChildResourceManager):
+    """CosmosDB Collection Resource
+
+    :example:
+    This policy will find all collections with Offer Throughput > 100
+
+    .. code-block:: yaml
+
+        policies:
+          - name: cosmosdb-high-throughput
+            resource: azure.cosmosdb-collection
+            filters:
+              - type: offer
+                key: content.offerThroughput
+                op: gt
+                value: 100
+
+    """
+    class resource_type(ChildTypeInfo):
+        parent_spec = ('cosmosdb', True)
+        parent_manager_name = 'cosmosdb'
+        raise_on_exception = False
+        annotate_parent = True
+
+    def enumerate_resources(self, parent_resource, type_info, **params):
+        data_client = CosmosDB.get_data_client(parent_resource, self)
+
+        try:
+            databases = list(data_client.ReadDatabases())
+        except HTTPFailure as e:
+            if e.status_code == 403:
+                log.error("Identity does not have permissions to read CosmosDB accounts"
+                          " or firewall is blocking access.")
+            else:
+                print(e)
+            return None
+
+        collections = []
+
+        for d in databases:
+            container_result = list(data_client.ReadContainers(d['_self']))
+            for c in container_result:
+                c.update({'c7n:document-endpoint':
+                         parent_resource.get('properties').get('documentEndpoint')})
+                collections.append(c)
+
+        return collections
+
+
+@CosmosDBCollection.filter_registry.register('offer')
+@CosmosDBDatabase.filter_registry.register('offer')
+class CosmosDBOfferFilter(ValueFilter):
+    schema = type_schema('offer', rinherit=ValueFilter.schema)
+    schema_alias = True
+
+    def process(self, resources, event=None):
+        futures = []
+        results = []
+
+        # Group all resources by account because offers are queried per account not per collection
+        account_sorted = sorted(resources, key=CosmosDBOfferFilter.account_key)
+        account_grouped = [list(it) for k, it in groupby(
+            account_sorted,
+            CosmosDBOfferFilter.account_key)]
+
+        # Process database groups in parallel
+        with self.executor_factory(max_workers=3) as w:
+            for resource_set in account_grouped:
+                futures.append(w.submit(self.process_resource_set, resource_set))
+
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.warning(
```

Comment:
ThreadHelper ? (And error messages :))

Reply:
I didn't think it was designed in a way to allow for the threading I was doing with custom grouping.